### PR TITLE
auto-patchelf: add support for __structuredAttrs

### PR DESCRIPTION
**Pull request by ConnorBaker**
## Description of changes

As discovered in #256324, `autoPatchelf` doesn't work correctly when `__structuredAttrs` is true. This patch fixes that.

This PR is tagged with a CUDA label because it being merged allows refactoring the work-around hack in #256324 where `autoPatchelfIgnoreMissingDeps` was moved inside `env` to ensure it was not treated as an array.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by jonringer — APPROVED** [review]
At first glance, this look right to me.

cc @jtojnar @infinisil @Mic92

**Comment by infinisil** [comment]
Can we have a test for this? :smiley:

**Comment by SomeoneSerge** [comment]
Oh, if you wanted to test expected failures, I recently played with running `nix-build` from inside a nixosTest: https://github.com/NixOS/nixpkgs/commit/4e6972e89034abcb67f5b19c0264821ddd9e2b0f

And expected successes don't even need a VM: you could e.g. generate a handful of derivations building dummy libraries. Then you pass them as buildInputs to another `__structuredAttrs` dummy derivation, which would build something linking dynamically the libraries from the previous step plus a few sonames to be ignored, strip their runpaths clean, and finally check that the hook has recovered the dummies and didn't fail at the missing SOs

**Comment by infinisil** [comment]
Yeah, see `pkgs/test/default.nix`, which is just recursive set of pakckages that get built by Hydra.

**Comment by jonringer** [comment]
> ran into this issue earlier... I'll try to find what package it was and get a minimal repo.

Turns out I had a different issue with `autoPatchelfHook`...

File: pkgs/build-support/setup-hooks/auto-patchelf.sh [hunk 1]
```diff
@@ -53,17 +53,25 @@ autoPatchelf() {
...
+        local runtimeDependenciesArray=( "${runtimeDependencies[@]}" )
+        local patchelfFlagsArray=( "${patchelfFlags[@]}" )
+    else
+        readarray -td' ' ignoreMissingDepsArray < <(echo -n "$autoPatchelfIgnoreMissingDeps")
+        local appendRunpathsArray=($appendRunpaths)
+        local runtimeDependenciesArray=($runtimeDependencies)
+        local patchelfFlagsArray=($patchelfFlags)
+    fi
+
     if [ "$autoPatchelfIgnoreMissingDeps" == "1" ]; then
```
[the remark this answers is not in the document]
Comment (ConnorBaker) [reply pkgs/build-support/setup-hooks/auto-patchelf.sh:68]: Woah! Sharp eyes @jtojnar! Let me see if I can come up with something a bit cleaner -- by this point `ignoreMissingDepsArray` is an array, so it should be enough to check for existence of `1`.

[the remark this answers is not in the document]
Comment (ConnorBaker) [reply pkgs/build-support/setup-hooks/auto-patchelf.sh]: Check out 45901c42fc91c1d1d46a811b9578b4d92da98135 and give me your thoughts on that -- it has the added benefit of now correctly issuing a deprecation warning when the ignore all option is used but not the head of the array.

**Comment by ConnorBaker** [comment]
I'm stuck :(

I don't have more time to devote to figuring out how to make tests for this. I've spent on the order of about eight hours struggling to figure out how to get this working (using https://github.com/NixOS/nixpkgs/pull/272752#issuecomment-1847284167 as a guide), but wasn't able to produce anything which even evaluated.

I agree that there should be tests for this, but given that there aren't any already I can use as a reference, I don't think I have the skill or time necessary to write them.

Is the lack of tests a blocker for merging?

**Comment by infinisil** [comment]
Ah sorry no I don't think a VM test is necessary. We can just use a Nix derivation that uses both `__structuredAttrs` and `autoPatchelfHook`.

- Get some binary that needs `autoPatchelfHook`. Probably easiest to just search for an existing package in Nixpkgs. I found `sslmate-agent` to be a good one, since it has a small closure.
- Create `pkgs/test/auto-patchelf-hook` and add a copy of `sslmate-agent`s Nix expression, but remove everything that's not necessary for the test, and add some `installCheckPhase` that verifies that the binary was patched successfully (you can use `patchelf --print...` for this).
- Then create another copy of that expression, but also set `__structuredAttrs = true`.
- Then create a `pkgs/test/auto-patchelf-hook/default.nix` that references both of those tests.
- Finally, include that `default.nix` from `pkgs/test/default.nix` with a `recurseIntoAttrs` like `autoPatchelfHook = recurseIntoAttrs (callPackages ./auto-patchelf-hook { })`
- Make sure both build successfully when you do `nix-build -A tests.autoPatchelfHook` (from the Nixpkgs root), ideally also do a quick check that when you break something intentionally, the check fails.

**Comment by yannham** [comment]
@infinisil thanks a lot for the detailed instructions! I set up something in [e92f2dc](https://github.com/NixOS/nixpkgs/pull/272752/commits/e92f2dc0d80a3a6dba33039390621cc4799369b5) following your advice. I built `sslmate-agent` with and without `autoPatchelf` to see the difference, and it seems that the only thing which is modified is the interpreter (no `RUNPATH` is set), so I wrote the test in consequence.

If I remove the `autoPatchelf` from the native build inputs of the test, and the test fails as expected, so at least it checks that patching happens. However, if I mess a bit with `autoPatchelf` (for example reverting back to before this PR) the test continues to pass, so I'm not entirely convinced this really triggers the code path which checks this new functionality.

**Comment by infinisil** [comment]
Oh I see, that's not a great test case then, because most code paths aren't actually exercised. `autoPatchelfHook` is most useful when providing libraries as dependencies.

I looked around a bit more and found `tonelib-jam` to be a better example. Here's my recommendation that also sets some `autoPatchelfHook` flags to influence its behavior:

```nix
{ stdenv
, fetchurl
, autoPatchelfHook
, dpkg
, freetype
}:

stdenv.mkDerivation {
  name = "auto-patchelf-test";

  src = fetchurl {
    url = "https://tonelib.net/download/221222/ToneLib-Jam-amd64.deb";
    sha256 = "sha256-c6At2lRPngQPpE7O+VY/Hsfw+QfIb3COIuHfbqqIEuM=";
  };

  unpackCmd = "dpkg -x $curSrc source";

  nativeBuildInputs = [
    dpkg
    autoPatchelfHook
  ];

  installPhase = ''
    mv usr $out
  '';

  buildInputs = [
    freetype
  ];

  autoPatchelfIgnoreMissingDeps = [
    "libGL.so.1"
    "libasound.so.2"
  ];

  runtimeDependencies = [
    "/some/dep"
    # This might only work with __structuredAttrs!
    "/a dep with a space"
  ];
}
```

You'll have to add an `installCheckPhase` still, but now you can verify both the result of `patchelf --print-interpreter` and `patchelf --print-rpath`.

**Comment by yannham** [comment]
@infinisil done in the last commit (update via force push). This test looks more solid :slightly_smiling_face: and indeed the test fails if I add the path with spaces when `__structureAttrs` is set to `false`.

File: pkgs/test/auto-patchelf-hook/package.nix [hunk 2]
```diff
@@ -0,0 +1,96 @@
+# This is a test for autoPatchelfHook. To test it, we just need a simple binary
+# which uses the hook. We picked took the derivation from tonelib-jam, which
```
Comment (yannham) [inline pkgs/test/auto-patchelf-hook/package.nix:2]:
```suggestion
# which uses the hook. We took the derivation from tonelib-jam, which
```

**Comment by ConnorBaker** [comment]
OfBorg failure (https://gist.github.com/GrahamcOfBorg/06df46e68204b4ce5b6d11e05b9fe30f) is unrelated and caused by the python package `aioconsole`.

Thank you @yannham for writing the test cases and dragging this over the finish line!

**Comment by vcunat** [comment]
Is this really safe for 23.11, i.e. sufficiently compatible? (PR #275923)

**Pull request merged**